src/libcrun/linux.c

```diff
@@ -5841,3 +5854,181 @@ libcrun_safe_chdir (const char *path, libcrun_error_t *err)
     }
   return 0;
 }
+
+static int
+run_in_container_namespace (libcrun_container_status_t *status, int (*callback) (void *, libcrun_error_t *), void *arg, libcrun_error_t *err)
+{
+  cleanup_close int pidfd = -1;
+  pid_t pid = status->pid;
+  int wait_status = 0;
+  int ret;
+
+  pidfd = syscall_pidfd_open (pid, 0);
+  if (UNLIKELY (pidfd < 0))
+    return crun_make_error (err, errno, "pidfd_open");
+
+  /* Check if the container is still running after opening the pidfd.  */
+  ret = libcrun_is_container_running (status, err);
+  if (UNLIKELY (ret < 0))
+    return ret;
+
+  if (ret == 0)
+    return crun_make_error (err, 0, "container not running");
+
+  /* must be vfork to propagate the error from the child proc.  */
+  pid = vfork ();
+  if (UNLIKELY (pid < 0))
+    return crun_make_error (err, errno, "vfork");
+
+  if (pid == 0)
+    {
+      ret = setns (pidfd, CLONE_NEWNS);
+      if (UNLIKELY (ret < 0))
+        {
+          crun_make_error (err, 0, "setns to target pid");
+          _exit (ret);
+        }
+      ret = chdir ("/");
+      if (UNLIKELY (ret < 0))
+        {
+          crun_make_error (err, errno, "chdir to `/`");
+          _exit (ret);
+        }
+
+      ret = callback (arg, err);
+      _exit (ret);
+    }
+  ret = waitpid_ignore_stopped (pid, &wait_status, 0);
+  if (UNLIKELY (ret < 0))
+    return ret;
+
+  return get_process_exit_status (wait_status);
```

Comment:
Is there a need for `crun_make_error` after `waitpid_ignore_stopped` and `get_process_exit_status`?

(I'm not sure `get_process_exit_status()` could ever return `-1`)

Reply:
Aha, I think I understand more now how it works.

But what about this scenario, could this ever happen?

At this source code
https://github.com/containers/crun/blob/182ccfce9cdf3d95cf6620b14c164a6439431a51/src/libcrun/linux.c#L5898
`callback()` returns `0`. No error was created.

and

At this source code
https://github.com/containers/crun/blob/182ccfce9cdf3d95cf6620b14c164a6439431a51/src/libcrun/utils.h#L411-L412
`waitpid_ignore_stopped()` returns `-1`

If that is possible, then `run_in_container_namespace()` returns `-1` without having an error created.

Reply:
yes thanks.  That can happen, I need to make sure there is an error created when `waitpid_ignore_stopped()` returns `< 0`;